### PR TITLE
Resampling indicator

### LINKED ISSUE #867: ResamplingIndicator
# Proposition for a restructuring of the `Indicator` class

- Create a new `ResamplingIndicator` for indicators that resample the data (most of them), those which have a `freq` argument.
- Remove all resampling-related operations from the base `Indicator` class.  This includes : missing values handling, "allowed_periods" check and handling of the `indexer` kwargs.
- The latter is currently only implemented on a few indicators, but we could generalize its use to _all_ resampling indicator. Thus, for all indicator when it makes sense, user could pass things like: `season='DJF'`. This would resolve an issue that pops up often in different scenarios : how to reduce the computation load without making `check_daily` fail.

### LINKED ISSUE #898: Allowed input freq as a parameter
Proposition to have indicators' datachecks using the class attribute `freq` when checking input's frequency.

By this I mean scraping the class-inheritance pattern of `Daily`, `Hourly` and `DailyWeeklyMonthy`. This will allow for a finer control over the accepted input freq for specific indicators. 

For example, it could allow whole modules, like `icclim` to allow weekly and monthly data. 

The default would stay `freq = 'D'`, but could be changed from indicator to indicator without having to inherit from a new base class. Ex:

```
from xclim.core.indicator import registry

tg_max_bis = registry['TG_MAX'](
    identifier="tg_max2",
    long_name="Maximum monthly mean temperature",
    description="{freq} maximum of monthly mean temperature.",
    cell_methods="time: mean within days time: maximum over months",
)
```
This creates a new `tg_max_bis` indicator that only accepts monthly input. The previous way was:
```
tg_max_bis = Monthly(
    identifier="tg_max",
    units="K",
    standard_name="air_temperature",
    long_name="Maximum monthly mean temperature",
    description="{freq} maximum of monthly mean temperature.",
    cell_methods="time: mean within days time: maximum over months",
    compute=indices.tg_max,
)
```
But there is no such thing as `Monthly` it would have to be created.

On a side note, this proposition removes _all_  conventional class-inheritance pattern from the creation of indicators. From now on, the only way to create a new indicator is through initialization of an existing class.

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #867 and fixes #898
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
  - [x] Link to issue (:issue:`number`) and pull request (:pull:`number`) has been added.
- [ ] `bumpversion patch` has been called on this branch
- [x] The relevant author information has been added to `.zenodo.json`

### What kind of change does this PR introduce?

* Split all resampling-related functionalities of the indicators into a new `ResamplingIndicator` subclass. I did this by creating private pre- and post-processing methods that are called in `__call__`. This makes it easy to "inject" functionalities in the indicator's processing. The "resampling-related functionalities" are : checking for the allowed "freq" arg and handling of missing values.
* Have the data input frequency expected by the indicator stored in a class attribute. This makes it easy to modify it when creating indicators. This improves the changes in #885.

### Does this PR introduce a breaking change?
I don't think so.

### Other information:

I think that with this change it becomes a bit clearer what are the uses cases for conventional subclassing of the `Indicator` class:
1. Add functionality by overriding one of the pre- or post-process methods.
2. Convenient way of modifying the following attributes : `context`, `src_freq`, `realm`, `title`, `abstract`, `keywords`, `references`, `notes` and `allowed_periods`. 
In both cases, the goal is to create a new _base_ class and _then_ to create multiple indicators.

I'll put this in a "developer's note" section of the doc in a further PR.

 This is also a first step towards a way to inject the `indexer` kwargs as planned in #899.

### CODE REVIEW

**Comment by aulemahal** [comment]
@bzah With this PR, I modified the way one can specifiy which input frequencies are accepted. See the `data/anuclim.yml` for an example. I hope it doesn't break your code too much!

The good part is that one can now inherit from an indicator created from `Daily` and still modifiy `src_freq`. `Daily` is only a convenience class, it now doesn't hardcode the frequency check, like it did before.

**Comment by bzah** [comment]
Thank you for your concern (and for the feature!) @aulemahal, it won't break my code I'm using atmos daily indicators directly and I have simply disabled the daily check for now on icc v5.
After this PR I might inherit them and change `src_freq`, but there is an ongoing discussion to see which indices/indicators should accept other frequencies.

File: xclim/core/indicator.py [hunk 1]
```diff
@@ -1285,47 +1232,109 @@ def injected_parameters(self):
...
+        self._missing = kls.execute
+        if self.missing_options:
+            kls.validate(**self.missing_options)
 
+        super().__init__(**kwds)
 
-class DailyWeeklyMonthly(Indicator):
-    """Indicator defined for inputs at daily, weekly or monthly frequencies.
+    def _preprocess_and_checks(self, das, params, indexer):
+        """Perform parent's checks and also checks if freq is allowed."""
```
Comment (huard) [inline xclim/core/indicator.py:1282]:
```suggestion
        """Perform parent's checks and also check if freq is allowed."""
```

**Review by huard — APPROVED** [review]
Good for me.

**Pull request merged**